flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java

```diff
@@ -0,0 +1,672 @@
+import org.apache.iceberg.flink.FlinkSchemaUtil;
+import org.apache.iceberg.flink.FlinkWriteConf;
+import org.apache.iceberg.flink.FlinkWriteOptions;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.flink.sink.committer.FilesCommittable;
+import org.apache.iceberg.flink.sink.committer.FilesCommittableSerializer;
+import org.apache.iceberg.flink.sink.committer.FilesCommitter;
+import org.apache.iceberg.flink.sink.writer.StreamWriter;
+import org.apache.iceberg.flink.sink.writer.StreamWriterState;
+import org.apache.iceberg.flink.sink.writer.StreamWriterStateSerializer;
+import org.apache.iceberg.flink.util.FlinkCompatibilityUtil;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.TypeUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Flink v2 sink offer different hooks to insert custom topologies into the sink. It includes
+ * writers that can write data to files in parallel and route commit info globally to one Committer.
+ * Post commit topology will take of compacting the already written files and updating the file log
+ * after the compaction.
+ *
+ * <pre>{@code
+ *                            Flink sink
+ *               +------------------------------------------------------------------+
+ *               |                                                                  |
+ * +-------+     | +-------------+       +---------------+                          |
+ * | Map 1 | ==> | | writer 1 | =|       | committer 1   |                          |
+ * +-------+     | +-------------+       +---------------+                          |
+ *               |                 \                                                |
+ *               |                  \                                               |
+ *               |                   \                                              |
+ * +-------+     | +-------------+    \  +---------------+        +---------------+ |
+ * | Map 2 | ==> | | writer 2 | =| ---  >| committer 2   |  --->  | post commit   | |
+ * +-------+     | +-------------+       +---------------+        +---------------+ |
+ *               |                                                                  |
+ *               +------------------------------------------------------------------+
+ * }</pre>
+ */
+public class IcebergSink
+    implements StatefulSink<RowData, StreamWriterState>,
+        WithPreWriteTopology<RowData>,
+        WithPreCommitTopology<RowData, FilesCommittable>,
+        WithPostCommitTopology<RowData, FilesCommittable> {
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergSink.class);
+
+  private final TableLoader tableLoader;
+  private final Table table;
```

Comment:
sink needs to be serializable. Iceberg `Table` is not serializable. There is a read-only `SerializableTable`.

Does this work in a distributed runtime? wondering if I miss sth.

Reply:
looking at the usage of the `table` object. This probably should be defined as a `SerializableTable`.

Reply:
The implementation classes of `Table` are all serializable.

Reply:
since we are using the `Table` interface here, it is better not to assume implementation class are serializable. Let's just use `SerializableTable` interface which is introduced for this purpose. As long as we make sure it is used for read-only path. Once `table` is serialized (when shipping to taskmanager), I believe it becomes a `SerializableTable` anyway.

Reply:
https://github.com/apache/iceberg/blob/be11b2f3b6441ddbddc694bfe24fd307bec0a808/flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSink.java#L644-L655
Thank you. When assigning the value of table, I changed it to this way. The return type of `SerializableTable.copyOfSerializableTable.copyOf()` is `Table`.